packages/grafana-ui/src/types/panelOptions.tsx

```diff
@@ -0,0 +1,181 @@
+   */
+  groupOptions: K;
+
+  /**
+   * Represents panel options that belong to a given layout group
+   */
+  options: Array<OptionUIModel<any, any> | SelectOptionUIModel<any, any>>;
+}
+
+/**
+ * Represents a container for a group of options.
+ * Panel UI for options is characterised by a title with a gradient background
+ * Use it for grouping related options:
+ *
+ *      ++++++++++++++++++++++++++++++++++++++
+ *      + Panel title                        +
+ *      ++++++++++++++++++++++++++++++++++++++
+ *      + Option 1                    on/off +
+ *      + Option 2                    on/off +
+ *      + ...                                +
+ *      + ...                                +
+ *      ++++++++++++++++++++++++++++++++++++++
+ *
+ */
+export interface PanelUIModel extends GroupLayoutUIModel<{ title: string }> {}
+
+/**
+ * Represents a group of options
+ * Fieldset UI does not have any characteristic UI elements. It's basically a group of options
+ * with a title:
+ *
+ *      ++++++++++++++++++++++++++++++++++++++
+ *      * Fieldset title                     *
+ *      *                                    *
+ *      + Option 1                    on/off +
+ *      + Option 2                    on/off +
+ *      + ...                                +
+ *      ++++++++++++++++++++++++++++++++++++++
+ */
+export interface FieldsetUIModel
+  extends GroupLayoutUIModel<{
+    legend: string;
+  }> {} // legend property name inspired by https://www.w3schools.com/tags/tag_legend.asp
+
+/**
+ * Below are typeguards used by PanelOptionsBuilder components
+ */
+export function isOptionsUIModel(model: any): model is OptionsUIModel<any> {
+  return (model as OptionsUIModel<any>).rows !== undefined;
+}
+export function isOptionModel(
+  option: OptionUIModel<any, any> | GroupLayoutUIModel<any>
+): option is OptionUIModel<any, any> {
+  return (option as OptionUIModel<any, any>).path !== undefined;
+}
+
+export function isOptionsPanelModel(ui: PanelUIModel | FieldsetUIModel): ui is PanelUIModel {
+  return (ui as PanelUIModel).groupOptions.title !== undefined;
+}
+
+export function isSelectOption(
+  option: OptionUIModel<any, any> | SelectOptionUIModel<any, any>
+): option is SelectOptionUIModel<any, any> {
+  return (option as SelectOptionUIModel<any, any>).options !== undefined;
+}
+
+// // const uiOptionModel = new UIModel<GaugeOptions>()
+// //   .addRow(/*{row config }*/)
+// //     .addColumn()
+// //       .addPanel('Basic settings') // Creates
+// //         .useStatsPicker('Show', 'valueMappings.stats')
+// //         .useUnitPicker('Unit', 'valueMappings.unit')
+// //         .useDecimalsInput('Decimals', 'valueMappings.decimals')
+// //         .useTextInput('Prefix', 'valueMappings.prefix')
+// //         .useTextInput('Suffix', 'valueMappings.suffix')
+// //         .addBooleanOption('Title', 'option.path')
+// //         .addSelectOption('Title', 'option.path', )
+// //       .endPanel()
```

Comment:
What are thoughts on how an external plugin would add to this?

Reply:
there will be a way to add custom components of course. Adding high level structures (groups) that can be reused will be a bit trickier :)

Reply:
Yes, I was thinking about it and my idea was to stick with the row/column approach, but enable used component config, i.e. `addRow({renderAs: MyRowComponent})`. Same with groupings: for now I thought only on panel(which name we need to change) and fiieldsets which are panel options specific. But the same approach could be taken towards grouping, i.e. current `addPanel` and `addFieldset` could use more generic function `addGroup({renderAs: PanelOptionsGroup})`.